Images on axton.gallery disappear momentarily when clicking on them

shallawa opened the pull request:
#### 41e6b037fd006ed630cd9114706ef9a803adbe2a
<pre>
Images on axton.gallery disappear momentarily when clicking on them
<a href="https://bugs.webkit.org/show_bug.cgi?id=298540">https://bugs.webkit.org/show_bug.cgi?id=298540</a>
<a href="https://rdar.apple.com/154363710">rdar://154363710</a>

Reviewed by Simon Fraser.

Currently if primitiveUnits is equal to userSpaceOnUse, primitiveSubRegion is
resolved against the viewport of the &lt;filter&gt; element&apos;s ancestor &lt;svg&gt;. Similarly
if filterUnits is to userSpaceOnUse, the filterRegion is resolved against the 
viewport of the &lt;filter&gt; element&apos;s ancestor &lt;svg&gt;.

The specs state that:

[1] Any length values within the filter definitions represent values in the current
local coordinate system in place at the time when the filter element is referenced.

[2] The local coordinate system is the coordinate system that is currently active
and which is used to define how coordinates and lengths are located and computed,
respectively, on the current canvas.

So the fix is:

1. For an HTML target element, its bounding box should be used to resolve
   primitiveSubRegion and filterRegion.

2. For an SVG target element, the viewport of its a ancestor &lt;svg&gt; should be
   used to resolve primitiveSubRegion and filterRegion.

[1] <a href="https://drafts.fxtf.org/filter-effects/#element-attrdef-filter-primitiveunits">https://drafts.fxtf.org/filter-effects/#element-attrdef-filter-primitiveunits</a>
[2] <a href="https://drafts.csswg.org/css-transforms-1/#local-coordinate-system">https://drafts.csswg.org/css-transforms-1/#local-coordinate-system</a>

Tests: imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space.html
       imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space.html

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/reference/svg-filter-filter-units-user-space-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/reference/svg-filter-primitive-units-user-space-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space.html: Added.
* Source/WebCore/rendering/CSSFilterRenderer.cpp:
(WebCore::createReferenceFilter):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::applyResource):
(WebCore::LegacyRenderSVGResourceFilter::resourceBoundingBox):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceMasker.cpp:
(WebCore::LegacyRenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::SVGLengthContext):
(WebCore::SVGLengthContext::resolveRectangle):
(WebCore::SVGLengthContext::viewportSize const):
* Source/WebCore/svg/SVGLengthContext.h:
(WebCore::SVGLengthContext::resolveRectangle):
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.cpp:
(WebCore::SVGFilterRenderer::create):
(WebCore::buildFilterEffectGraph):
(WebCore::SVGFilterRenderer::buildExpression):
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.h:

Canonical link: <a href="https://commits.webkit.org/301424@main">https://commits.webkit.org/301424@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/034ae413cf8d184ca892fa06c6e1941e5538568d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36399 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/132829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/77820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/536438e0-eac2-48ff-a73e-7453c1d6ddab) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54183 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/132829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/77820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8a090c49-b019-43e2-ad72-c0c2597ecaa5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128919 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/37042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/112658 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/132829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8ac02235-a85e-4bf8-8000-f2cdf874b247) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/30844 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/76300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/106824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31058 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/135517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/52748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40479 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/135517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108872 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/135517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26516 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49539 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27871 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52642 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58459 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/51985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/55333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/53682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -59,7 +59,7 @@ SVGLengthContext::~SVGLengthContext() = default;
 {
     ASSERT(type != SVGUnitTypes::SVG_UNIT_TYPE_UNKNOWN);
-    if (type != SVGUnitTypes::SVG_UNIT_TYPE_USERSPACEONUSE) {
+    if (!context || type != SVGUnitTypes::SVG_UNIT_TYPE_USERSPACEONUSE) {
```
I would prefer if we would null-check context at the call site and pass in a const SVGElement reference instead.

shallawa replied:
This approach was wrong. Treating all the lengths always as percentages if `context` is null was wrong. I fixed this by making `SVGLengthContext::resolveRectangle()` pass `viewport.size()` to the constructor of `SVGLengthContext`. `SVGLengthContext::viewportSize()` will return `m_viewportSize` if  `context` is null.

shallawa replied:
With the old approach a primitive like this:
```
    <feColorMatrix width="5" height="5"></fecolormatrix>
```
Will be resolved as `width="500%"` and `height="500%"` if the containing filter is referenced from a none `SVGElement`.

nikolaszimmermann reviewed (commented):
Thanks Said, looks good. Some comments:
Do you plan to contribute those tests to WPT?

smfr commented on Source/WebCore/rendering/CSSFilterRenderer.cpp:
```diff
@@ -235,9 +235,14 @@ static RefPtr<SVGFilterRenderer> createReferenceFilter(CSSFilterRenderer& filter
         return nullptr;
 
-    auto filterRegion = SVGLengthContext::resolveRectangle<SVGFilterElement>(filterElement.get(), filterElement->filterUnits(), targetBoundingBox);
+    RefPtr targetElement = renderer.element();
```
No need for this local variable.

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
```diff
@@ -46,7 +46,15 @@ RenderSVGResourceFilter::~RenderSVGResourceFilter() = default;
+    if (!renderer)
+        return SVGLengthContext::resolveRectangle(filterElement.get(), filterElement->filterUnits(), object.objectBoundingBox());
+
+    RefPtr targetElement = renderer->element();
```
Ditto.

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
```diff
@@ -100,9 +101,12 @@ auto LegacyRenderSVGResourceFilter::applyResource(RenderElement& renderer, const
-    auto targetBoundingBox = renderer.objectBoundingBox();
     Ref filterElement = this->filterElement();
-    auto filterRegion = SVGLengthContext::resolveRectangle<SVGFilterElement>(filterElement.ptr(), filterElement->filterUnits(), targetBoundingBox);
+    RefPtr targetElement = renderer.element();
```
Ditto.

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
```diff
@@ -221,7 +225,15 @@ void LegacyRenderSVGResourceFilter::postApplyResource(RenderElement& renderer, G
+    if (!renderer)
+        return SVGLengthContext::resolveRectangle(filterElement.get(), filterElement->filterUnits(), object.objectBoundingBox());
+
+    RefPtr targetElement = renderer->element();
```
Ditto

smfr commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -50,8 +50,9 @@
 namespace WebCore {
 
-SVGLengthContext::SVGLengthContext(const SVGElement* context)
+SVGLengthContext::SVGLengthContext(const SVGElement* context, const std::optional<FloatSize>& viewportSize)
```
The alternative here is to allow the creation of a `SVGLengthContext` with a CSS renderer; did you consider that?

shallawa committed:
Images on axton.gallery disappear momentarily when clicking on them
https://bugs.webkit.org/show_bug.cgi?id=298540
rdar://154363710

Reviewed by Simon Fraser.

Currently if primitiveUnits is equal to userSpaceOnUse, primitiveSubRegion is
resolved against the viewport of the <filter> element's ancestor <svg>. Similarly
if filterUnits is to userSpaceOnUse the filterRegion and the is resolved against
the viewport of the <filter> element's ancestor <svg>.

The specs state that:

[1] Any length values within the filter definitions represent values in the current
local coordinate system in place at the time when the filter element is referenced.

[2] The local coordinate system is the coordinate system that is currently active
and which is used to define how coordinates and lengths are located and computed,
respectively, on the current canvas.

So the fix is:

1. For an HTML target element, its bounding boxes should be used to resolve
   primitiveSubRegion and filterRegion.

2. For an SVG target element, the viewport of its a ancestor <svg> should be
   used to resolve primitiveSubRegion and filterRegion.

[1] https://drafts.fxtf.org/filter-effects/#element-attrdef-filter-primitiveunits
[2] https://drafts.csswg.org/css-transforms-1/#local-coordinate-system

Tests: imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space.html
       imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space.html

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/reference/svg-filter-filter-units-user-space-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/reference/svg-filter-primitive-units-user-space-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-filter-units-user-space.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/svg-filter-primitive-units-user-space.html: Added.
* Source/WebCore/rendering/CSSFilterRenderer.cpp:
(WebCore::createReferenceFilter):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::applyResource):
(WebCore::LegacyRenderSVGResourceFilter::resourceBoundingBox):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceMasker.cpp:
(WebCore::LegacyRenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::SVGLengthContext):
(WebCore::SVGLengthContext::resolveRectangle):
(WebCore::SVGLengthContext::viewportSize const):
* Source/WebCore/svg/SVGLengthContext.h:
(WebCore::SVGLengthContext::resolveRectangle):
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.cpp:
(WebCore::SVGFilterRenderer::create):
(WebCore::buildFilterEffectGraph):
(WebCore::SVGFilterRenderer::buildExpression):
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.h:

Canonical link: https://commits.webkit.org/301424@main
